### PR TITLE
Refactor models to use mapping

### PR DESCRIPTION
**Pull request by ajaxscape**
Refactor models to use mapping between dynamics and model data to simplify binding and unbinding of entity associations within dynamics.

### CODE REVIEW

File: src/models/base.model.js [hunk 1]
```diff
@@ -1,30 +1,100 @@
...
 module.exports = class BaseModel {
+  constructor (...args) {
+    const data = args.pop()
+    if (data) {
+      Object.assign(this, data)
+    }
+    Utilities.convertFromDynamics(this)
+  }
+
+  static selectedDynamicsFields (filter = () => true) {
```
Comment (duncannixon) [inline src/models/base.model.js]: This needs commenting to explain the purpose of the method and what the filter parameter does.

File: src/models/base.model.js [hunk 2]
```diff
@@ -1,30 +1,100 @@
...
+    return this.mapping()
+      .filter(filter)
+      .map(({dynamics}) => dynamics)
+  }
+
+  static dynamicsToModel (dynamicsData, filter = () => true) {
```
Comment (duncannixon) [inline src/models/base.model.js]: This needs commenting to explain the purpose of the method and what the parameters do.

File: src/models/base.model.js [hunk 3]
```diff
@@ -1,30 +1,100 @@
...
+      .filter(({defaultVal}) => !defaultVal)
+      .forEach(({field, dynamics}) => {
+        // set values in javascript objects by specifying a path eg 'dob.month'.
+        // if the path doesn't exist yet, it will be created.
+        ObjectPath.set(modelData, field, dynamicsData[dynamics])
+      })
+    return new this(modelData)
+  }
+
+  modelToDynamics (filter = () => true) {
```
Comment (duncannixon) [inline src/models/base.model.js]: This needs commenting to explain the purpose of the method and what the parameters do.

File: src/models/base.model.js [hunk 4]
```diff
@@ -1,30 +1,100 @@
...
+            if (this[field] !== undefined || field !== 'id') {
+              dynamicsData[dynamics] = ObjectPath.get(this, field)
+            }
+          }
+        }
+      })
+    return dynamicsData
+  }
+
   toString () {
```
Comment (duncannixon) [inline src/models/base.model.js]: Did this method actually need to be re-written?

File: src/models/base.model.js [hunk 5]
```diff
@@ -1,30 +1,100 @@
...
     // Class name
-    let returnValue = `${this.constructor.name}: {\n`
+    const className = this.constructor.name
```
Comment (duncannixon) [inline src/models/base.model.js]: Why did this method need to be re-written?

File: src/models/base.model.js [hunk 6]
```diff
@@ -1,30 +1,100 @@
...
 
-    return returnValue
+    return `${className}: ${properties}`
   }
 
   isNew () {
     return !this.id
   }
 
+  async _deleteBoundReferences (dynamicsDal) {
```
Comment (duncannixon) [inline src/models/base.model.js]: This needs commenting to explain the purpose of the method.

File: src/models/base.model.js (hunk 5 above)
Comment (ajaxscape) [reply src/models/base.model.js]: I changed the code to allow a toString of nested objects.

**Pull request merged**